### PR TITLE
New stat rocksdb.{cf|db}-write-stall-stats exposed in a structural way

### PR DESCRIPTION
**Pull request by hx235**
**Context/Summary:**
Users are interested in figuring out what has caused write stall. 
- Refactor write stall related stats from property `kCFStats` into its own db property `rocksdb.cf-write-stall-stats` as a map or string. For now, this only contains count of different combination of (CF-scope `WriteStallCause`) + (`WriteStallCondition`)
- Add new `WriteStallCause::kWriteBufferManagerLimit` to reflect write stall caused by write buffer manager
- Add new `rocksdb.db-write-stall-stats`. For now, this only contains `WriteStallCause::kWriteBufferManagerLimit` + `WriteStallCondition::kStopped`

- Expose functions in new class `WriteStallStatsMapKeys` for examining the above two properties returned as map
- Misc: rename/comment some write stall InternalStats for clarity

**Test:**
- New UT
- Stress test
`python3 tools/db_crashtest.py blackbox --simple --get_property_one_in=1`
- Perf test: Both converge very slowly at similar rates but post-change has higher average ops/sec than pre-change even though they are run at the same time.
```
./db_bench -seed=1679014417652004 -db=/dev/shm/testdb/ -statistics=false -benchmarks="fillseq[-X60]" -key_size=32 -value_size=512 -num=100000 -db_write_buffer_size=655 -target_file_size_base=655 -disable_auto_compactions=false -compression_type=none -bloom_bits=3
```
pre-change:
```
fillseq [AVG 15 runs] : 1176 (± 732) ops/sec;    0.6 (± 0.4) MB/sec
fillseq      :    1052.671 micros/op 949 ops/sec 105.267 seconds 100000 operations;    0.5 MB/s
fillseq [AVG 16 runs] : 1162 (± 685) ops/sec;    0.6 (± 0.4) MB/sec
fillseq      :    1387.330 micros/op 720 ops/sec 138.733 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 17 runs] : 1136 (± 646) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1232.011 micros/op 811 ops/sec 123.201 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 18 runs] : 1118 (± 610) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1282.567 micros/op 779 ops/sec 128.257 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 19 runs] : 1100 (± 578) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1914.336 micros/op 522 ops/sec 191.434 seconds 100000 operations;    0.3 MB/s
fillseq [AVG 20 runs] : 1071 (± 551) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1227.510 micros/op 814 ops/sec 122.751 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 21 runs] : 1059 (± 525) ops/sec;    0.5 (± 0.3) MB/sec
```
post-change:
```
fillseq [AVG 15 runs] : 1226 (± 732) ops/sec;    0.6 (± 0.4) MB/sec
fillseq      :    1323.825 micros/op 755 ops/sec 132.383 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 16 runs] : 1196 (± 687) ops/sec;    0.6 (± 0.4) MB/sec
fillseq      :    1223.905 micros/op 817 ops/sec 122.391 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 17 runs] : 1174 (± 647) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1168.996 micros/op 855 ops/sec 116.900 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 18 runs] : 1156 (± 611) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1348.729 micros/op 741 ops/sec 134.873 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 19 runs] : 1134 (± 579) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1196.887 micros/op 835 ops/sec 119.689 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 20 runs] : 1119 (± 550) ops/sec;    0.6 (± 0.3) MB/sec
fillseq      :    1193.697 micros/op 837 ops/sec 119.370 seconds 100000 operations;    0.4 MB/s
fillseq [AVG 21 runs] : 1106 (± 524) ops/sec;    0.6 (± 0.3) MB/sec
```

### CODE REVIEW

File: include/rocksdb/listener.h [hunk 1]
```diff
@@ -194,10 +194,32 @@ enum class BackgroundErrorReason {
...
+
 enum class WriteStallCondition {
-  kNormal,
   kDelayed,
   kStopped,
+  // Add new WriteStallCondition before `kNormal`
+  kNormal,
+};
+
+struct CFWriteStallStatsMapKeys {
```
Comment (hx235) [inline include/rocksdb/listener.h:213]: Would it be too much if we create a .h in include just for this class and WriteStallCause, WriteStallCondition?

File: include/rocksdb/listener.h [hunk 2]
```diff
@@ -194,10 +194,32 @@ enum class BackgroundErrorReason {
...
+  static const std::string& TotalStop();
+  static const std::string& TotalSlowDown();
+  static const std::string& L0FileCountLimitSlowdownsWithOngoingCompaction();
+  static const std::string& L0FileCountLimitStopsWithOngoingCompaction();
+
+  // REQUIRES: cause != `WriteStallCause::kNone`, condition !=
+  // `WriteStallCondition::kNormal`
+  static std::string CauseConditionCount(WriteStallCause cause,
+                                         WriteStallCondition condition);
 };
```
Comment (hx235) [inline include/rocksdb/listener.h:223]: Would it be too much if we create a .h in include just for this class and WriteStallCause, WriteStallCondition?

**Comment by hx235** [comment]
Fix an incorrect assertion added by me exposed in mini crash test

File: db/write_stall_stats.cc [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+WriteStallStatsMapKeys::CFL0FileCountLimitDelaysWithOngoingCompaction() {
+  static const std::string ret =
+      "cf-l0-file-count-limit-delays-with-ongoing-compaction";
+  return ret;
+}
+
+const std::string&
+WriteStallStatsMapKeys::CFL0FileCountLimitStopsWithOngoingCompaction() {
+  static const std::string ret =
+      "cf-l0-file-count-limit-stopps-with-ongoing-compaction";
```
Comment (ajkr) [inline db/write_stall_stats.cc:127]: s/stopps/stops/

File: include/rocksdb/types.h [hunk 4]
```diff
@@ -63,4 +63,40 @@ enum EntryType {
...
+  kNone,
+};
+
+constexpr uint32_t kNumCFScopeWriteStallCauses =
+    static_cast<uint32_t>(WriteStallCause::kCFScopeWriteStallCauseEnumMax) -
+    static_cast<uint32_t>(WriteStallCause::kMemtableLimit);
+
+constexpr uint32_t kNumDBScopeWriteStallCauses =
+    static_cast<uint32_t>(WriteStallCause::kDBScopeWriteStallCauseEnumMax) -
+    static_cast<uint32_t>(WriteStallCause::kWriteBufferManagerLimit);
```
Comment (ajkr) [inline include/rocksdb/types.h:93]: These seem mainly for your convenience. Can they be moved to one of the internal files?

**Review by ajkr — CHANGES_REQUESTED** [review]
Great progress! Left a few suggestions, mainly around reporting a bit less info in the DB variant of the new property

File: db/db_properties_test.cc [hunk 5]
```diff
@@ -2106,6 +2106,114 @@ TEST_F(DBPropertiesTest, GetMapPropertyBlockCacheEntryStats) {
   ASSERT_EQ(3 * kNumCacheEntryRoles + 4, values.size());
 }
 
+TEST_F(DBPropertiesTest, GetMapPropertyWriteStallStats) {
```
Comment (hx235) [inline db/db_properties_test.cc:2109]: TODO: adding a seperate UT to verify format of writes stall cause/condition/internal stats enum and hardcoded map in write_stall_stats.cc

File: HISTORY.md [hunk 6]
```diff
@@ -14,6 +14,8 @@
 * Add statistics rocksdb.secondary.cache.filter.hits, rocksdb.secondary.cache.index.hits, and rocksdb.secondary.cache.filter.hits
 * Added a new PerfContext counter `internal_merge_point_lookup_count` which tracks the number of Merge operands applied while serving point lookup queries.
 * Add new statistics rocksdb.table.open.prefetch.tail.read.bytes, rocksdb.table.open.prefetch.tail.{miss|hit}
+* Add new db properties `rocksdb.cf-write-stall-stats`, `rocksdb.db-write-stall-stats`and APIs to examine them in a structured way.
+In particular, users of `GetMapProperty()` with property `kCFWriteStallStats`/`kDBWriteStallStats` can now use the functions in `WriteStallStatsMapKeys` to find stats in the map.
```
[the remark this answers is not in the document]
Comment (hx235) [reply HISTORY.md]: Fixed

File: db/internal_stats.cc [hunk 7]
```diff
@@ -1589,6 +1637,74 @@ void InternalStats::DumpDBStats(std::string* value) {
...
+      if (condition == WriteStallCondition::kDelayed) {
+        total_delay += stat;
+      } else if (condition == WriteStallCondition::kStopped) {
+        total_stop += stat;
+      }
+    }
+  }
+
+  (*value)[WriteStallStatsMapKeys::TotalStop()] = std::to_string(total_stop);
+  (*value)[WriteStallStatsMapKeys::TotalDelay()] = std::to_string(total_delay);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.cc]: Fixed

File: db/internal_stats.h [hunk 8]
```diff
@@ -706,6 +746,8 @@ class InternalStats {
     // Total time writes delayed by stalls.
     uint64_t write_stall_micros;
     double seconds_up;
+    // Total counts of DB-scope write stalls
+    uint64_t stall_count;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.h]: Fixed

File: db/internal_stats.cc [hunk 9]
```diff
@@ -1589,6 +1637,74 @@ void InternalStats::DumpDBStats(std::string* value) {
...
+
+  if (total_stall_count) {
+    *total_stall_count =
+        std::stoi(write_stall_stats_map[WriteStallStatsMapKeys::TotalStop()]) +
+        std::stoi(write_stall_stats_map[WriteStallStatsMapKeys::TotalDelay()]);
+    if (*total_stall_count > 0) {
+      str << "interval: " << *total_stall_count - db_stats_snapshot_.stall_count
+          << " total count\n";
+    }
+  }
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.cc]: Fixed

File: db/internal_stats.cc [hunk 10]
```diff
@@ -1709,36 +1825,78 @@ void InternalStats::DumpCFMapStatsByPriority(
...
+  str << "Write Stall (count): ";
+
+  for (const auto& name_and_stat : write_stall_stats_map) {
+    str << name_and_stat.first << ": " << name_and_stat.second << ", ";
+  }
+
+  if (total_stall_count) {
+    *total_stall_count =
+        std::stoi(write_stall_stats_map[WriteStallStatsMapKeys::TotalStop()]) +
+        std::stoi(write_stall_stats_map[WriteStallStatsMapKeys::TotalDelay()]);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.cc]: Oh right - i forgot this was mainly for the total count and the hardcoded map

[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.cc]: Fixed by ParseUint64()

File: db/write_stall_stats.cc (hunk 3 above)
Comment (hx235) [reply db/write_stall_stats.cc]: Fixed

File: include/rocksdb/types.h (hunk 4 above)
Comment (hx235) [reply include/rocksdb/types.h]: Fixed

File: db/internal_stats.cc [hunk 11]
```diff
@@ -1472,6 +1513,8 @@ void InternalStats::DumpDBMapStats(
   }
   double seconds_up = (clock_->NowMicros() - started_at_) / kMicrosInSec;
   (*db_stats)["db.uptime"] = std::to_string(seconds_up);
+
+  DumpDBMapStatsWriteStall(db_stats);
```
Comment (ajkr) [inline db/internal_stats.cc:1517]: Is this redundant with the addition to db_stats_type_to_info?

File: db/internal_stats.cc [hunk 12]
```diff
@@ -1591,6 +1638,59 @@ void InternalStats::DumpDBStats(std::string* value) {
...
+      uint64_t stat =
+          db_stats_[static_cast<std::size_t>(internal_db_stat)].load(
+              std::memory_order_relaxed);
+      (*value)[name] = std::to_string(stat);
+
+      if (condition == WriteStallCondition::kDelayed) {
+        total_delay += stat;
+      } else if (condition == WriteStallCondition::kStopped) {
+        total_stop += stat;
+      }
```
Comment (ajkr) [inline db/internal_stats.cc]: It's unused

File: db/internal_stats.h [hunk 13]
```diff
@@ -104,15 +104,36 @@ class InternalStats {
...
+    L0_FILE_COUNT_LIMIT_DELAYS_WITH_ONGOING_COMPACTION,  // Write slowdown
+                                                         // caused by l0 file
+                                                         // count limit while
+                                                         // there is ongoing
+                                                         // L0 compaction
+    L0_FILE_COUNT_LIMIT_STOPS_WITH_ONGOING_COMPACTION,   // Write stop caused by
+                                                         // l0 file count limit
+                                                         // while there is
+                                                        // ongoing L0 compaction
+    // End of repetitive write stall stats
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/internal_stats.h]: Fixed

File: db/write_stall_stats.cc [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        "pending-compaction-bytes",
+        "invalid",
+        "write-buffer-manager-limit"
+        "invalid",
+    }};
+
+std::array<std::string, static_cast<uint32_t>(WriteStallCondition::kNormal)>
+    kWriteStallConditionToHyphenString{{
+        "delayed",
+        "stopped",
```
Comment (ajkr) [inline db/write_stall_stats.cc:25]: "stopped"

File: db/write_stall_stats.cc [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+      static_cast<uint32_t>(WriteStallCause::kDBScopeWriteStallCauseEnumMax) -
+      kNumDBScopeWriteStallCauses;
+  uint32_t upper_bound =
+      static_cast<uint32_t>(WriteStallCause::kDBScopeWriteStallCauseEnumMax) -
+      1;
+  return lower_bound <= int_cause && int_cause <= upper_bound;
+}
+
+const std::string& WriteStallStatsMapKeys::TotalStop() {
+  static const std::string kTotalStop = "total-stop";
```
Comment (ajkr) [inline db/write_stall_stats.cc:111]: "stop"

File: db/write_stall_stats.cc [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+WriteStallStatsMapKeys::CFL0FileCountLimitDelaysWithOngoingCompaction() {
+  static const std::string ret =
+      "cf-l0-file-count-limit-delays-with-ongoing-compaction";
+  return ret;
+}
+
+const std::string&
+WriteStallStatsMapKeys::CFL0FileCountLimitStopsWithOngoingCompaction() {
+  static const std::string ret =
+      "cf-l0-file-count-limit-stops-with-ongoing-compaction";
```
Comment (ajkr) [inline db/write_stall_stats.cc:130]: "stops"

The point of this and the above two comments isn't to increase my comment count with low effort, but to point out the terminology changes slightly throughout the file. Can we use "stops" in all the places (and "delays")?

**Review by ajkr — APPROVED** [review]
Looks great - thanks!!!

File: db/internal_stats.cc (hunk 12 above)
Comment (hx235) [reply db/internal_stats.cc]: ha - noticed that too - fixed

File: db/write_stall_stats.cc (hunk 16 above)
Comment (hx235) [reply db/write_stall_stats.cc:130]: sounds good to me!

File: db/internal_stats.cc (hunk 11 above)
Comment (hx235) [reply db/internal_stats.cc:1517]: Oh right we remove the sum stats in DumpDBMapStatsWriteStall now they are the same. Removed it here.

File: db/write_stall_stats.cc (hunk 16 above)
Comment (hx235) [reply db/write_stall_stats.cc:130]: Fixed

File: db/write_stall_stats.cc (hunk 15 above)
Comment (hx235) [reply db/write_stall_stats.cc]: Fixed

File: db/write_stall_stats.cc (hunk 14 above)
Comment (hx235) [reply db/write_stall_stats.cc]: Fixed

File: db/internal_stats.cc (hunk 11 above)
Comment (hx235) [reply db/internal_stats.cc]: Fixed

**Pull request closed without merging**